src/Workspaces/Core/Portable/FindSymbols/SymbolTree/SymbolTreeInfo_Metadata.cs

```diff
@@ -80,9 +80,27 @@ private static Metadata GetMetadataNoThrow(PortableExecutableReference reference
             var metadata = GetMetadataNoThrow(reference);
             if (metadata == null)
             {
-                return null;
+                return SpecializedTasks.Default<SymbolTreeInfo>();
+            }
+
+            // Try to acquire the data outside the lock.  That way we can avoid any sort of 
+            // allocations around acquiring the task for it.  Note: once ValueTask is available
+            // (and enabled in the language), we'd likely want to use it here. (Presuming 
+            // the lock is not being held most of the time).
+            SymbolTreeInfo info;
+            if (s_metadataIdToInfo.TryGetValue(metadata.Id, out info))
+            {
+                return Task.FromResult(info);
```

Comment:
Why not update `s_metadataIdToInfo` to store a `Task<SymbolTreeInfo>`, so you can give back the existing task every time.